File: app/src/main/java/de/kuschku/libquassel/primitives/serializers/BufferInfoSerializer.java
/*
 * QuasselDroid - Quassel client for Android
 * Copyright (C) 2016 Janne Koschinski
 * Copyright (C) 2016 Ken Børge Viktil
 * Copyright (C) 2016 Magnus Fjell
 * Copyright (C) 2016 Martin Sandsmark <martin.sandsmark@kde.org>
 *
 * This program is free software: you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as published by the Free
 * Software Foundation, either version 3 of the License, or (at your option)
 * any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package de.kuschku.libquassel.primitives.serializers;

import android.support.annotation.NonNull;

import java.io.IOException;
import java.nio.ByteBuffer;
import java.nio.channels.ByteChannel;

import de.kuschku.libquassel.primitives.types.BufferInfo;

public class BufferInfoSerializer implements PrimitiveSerializer<BufferInfo> {
    @NonNull
    private static final BufferInfoSerializer serializer = new BufferInfoSerializer();

    private BufferInfoSerializer() {
    }

    @NonNull
    public static BufferInfoSerializer get() {
        return serializer;
    }

    @Override
    public void serialize(@NonNull ByteChannel channel, @NonNull BufferInfo data) throws IOException {
        IntSerializer.get().serialize(channel, data.id);
        IntSerializer.get().serialize(channel, data.networkId);
        ShortSerializer.get().serialize(channel, data.type.id);
        IntSerializer.get().serialize(channel, data.groupId);
        ByteArraySerializer.get().serialize(channel, data.name);
    }

    @NonNull
    @Override
    public BufferInfo deserialize(@NonNull final ByteBuffer buffer) throws IOException {
        return BufferInfo.create(
                IntSerializer.get().deserialize(buffer),
                IntSerializer.get().deserialize(buffer),
                BufferInfo.Type.fromId(ShortSerializer.get().deserialize(buffer)),
                IntSerializer.get().deserialize(buffer),
                ByteArraySerializer.get().deserialize(buffer)
        );
    }
}

File: lib/src/main/java/de/kuschku/libquassel/protocol/primitive/serializer/BufferInfoSerializer.kt
/*
 * Quasseldroid - Quassel client for Android
 *
 * Copyright (c) 2020 Janne Mareike Koschinski
 * Copyright (c) 2020 The Quassel Project
 *
 * This program is free software: you can redistribute it and/or modify it
 * under the terms of the GNU General Public License version 3 as published
 * by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program. If not, see <http://www.gnu.org/licenses/>.
 */

package de.kuschku.libquassel.protocol.primitive.serializer

import de.kuschku.libquassel.protocol.Buffer_Type
import de.kuschku.libquassel.quassel.BufferInfo
import de.kuschku.libquassel.quassel.QuasselFeatures
import de.kuschku.libquassel.util.nio.ChainedByteBuffer
import java.nio.ByteBuffer

object BufferInfoSerializer : Serializer<BufferInfo> {
  override fun serialize(buffer: ChainedByteBuffer, data: BufferInfo, features: QuasselFeatures) {
    BufferIdSerializer.serialize(buffer, data.bufferId, features)
    NetworkIdSerializer.serialize(buffer, data.networkId, features)
    ShortSerializer.serialize(buffer, data.type.toShort(), features)
    IntSerializer.serialize(buffer, data.groupId, features)
    StringSerializer.UTF8.serialize(buffer, data.bufferName, features)
  }

  override fun deserialize(buffer: ByteBuffer, features: QuasselFeatures): BufferInfo {
    val bufferId = BufferIdSerializer.deserialize(buffer, features)
    val networkId = NetworkIdSerializer.deserialize(buffer, features)
    val type = Buffer_Type.of(ShortSerializer.deserialize(buffer, features))
    val groupId = IntSerializer.deserialize(buffer, features)
    val bufferName = StringSerializer.UTF8.deserialize(buffer, features)
    return BufferInfo(
      bufferId = bufferId,
      networkId = networkId,
      type = type,
      groupId = groupId,
      bufferName = bufferName
    )
  }
}
